### DISCUSSION TITLE
Running Hedy programs line-by-line and examining variables

### DISCUSSION THREAD

**Discussion by redelaar** [hedyorg/hedy#2704] [Announcements]
Since a couple of days it is possible to run Hedy programs line-by-line and to examine variables!

### Line by line execution
It is possible to execute Hedy programs line-by-line, using the debug `🐞 button`. This button can be found below the text editor and will execute the first line of the program. The user can execute the next lines by pressing on the forward `⏩ button`. To stop the line-by-line execution, press the `🛑 button`. To start the line-by-line execution over from line 1, press the `🔁 button`.

The line that is being executed will be highlighted in blue. In the image below, `add appel to taarten` is being executed. 

![image](https://user-images.githubusercontent.com/33595265/167370860-a623ce65-857c-4666-b5fc-39a6a3ff3018.png)

### Examining variables
The second feature is a tool that shows the self-created variables in a program and their values. By clicking on the blue `🏷 button` in the output area, the view will be expanded or minimized. All variables that users make in their program will be shown during/after the execution of their program. The values will change automatically if this happens in the program. For example, in the screen capture above, the variables "taarten" and its value can be seen. 

Both features can be used separately. It is completely possible to use the variable view without using the line-by-line execution at the same time.

**Comment by redelaar** [comment]
Tagging @Felienne so she can move this post to "announcements" :)

**Comment by MarleenGilsing** [comment]
I love these new features! Great job, @fRedelaar! 
I tried them out today and I do have one small remark about the debugger. Wouldn't it make more sense to put the next button on the right and the back button on the left? Like in any type of mediaplayer?

**Reply by redelaar** [reply]
Oh, that is quite clever :). So the order would go from `⏩🔁🛑` to `🔁🛑⏩`  @MarleenGilsing ?

**Reply by MarleenGilsing** [reply]
Yes I think that would make more sense for the user.

**Reply by redelaar** [reply]
Thanks for the suggestion! I created a pull request. #2733 :)

**Comment by MarleenGilsing** [comment]
Also, maybe you can change the variable name in level 1 to echo instead of answer?
Or just diasble the feature in level 1, if changing it is hard.
![image](https://user-images.githubusercontent.com/80678586/168052940-973ae441-afe3-4c59-bef2-e5ccfa5f6e1a.png)

**Reply by OnnoEbbens** [reply]
I think it is a good idea to disable this feature in level 1 because you don't have variables yet in level 1

**Reply by redelaar** [reply]
Thanks for the input both @OnnoEbbens  and @MarleenGilsing ! :) I have pushed a pull request right away. The variable view will be disabled in level 1.
